lib/async-emitter.js

```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+const { iter } = require('@metarhia/common');
+
+class AsyncEmitter {
+  constructor() {
+    this.events = new Map();
+    this.wrappers = new Map();
+  }
+
+  on(name, fn) {
+    let event = this.events.get(name);
+    if (!event) {
+      event = new Set();
+      this.events.set(name, event);
+    }
+    event.add(fn);
+  }
+
+  once(name, fn) {
+    if (fn === undefined) {
+      return new Promise(resolve => {
+        this.once(name, resolve);
+      });
+    }
+    const wrapper = (...args) => {
+      this.remove(name, wrapper);
+      return fn(...args);
+    };
+    this.wrappers.set(fn, wrapper);
+    this.on(name, wrapper);
+    return undefined;
+  }
+
+  async emit(name, ...args) {
+    const event = this.events.get(name);
+    if (!event) return undefined;
+    const listeners = event.values();
+    const promises = iter(listeners).map(fn => fn(...args));
+    return Promise.all(promises);
+  }
+
+  remove(name, fn) {
+    const event = this.events.get(name);
+    if (!event) return;
+    if (event.has(fn)) event.delete(fn);
+    const wrapper = this.wrappers.get(fn);
+    if (wrapper) {
+      this.wrappers.delete(fn);
+      event.delete(wrapper);
+    }
+    if (event.size === 0) this.events.delete(name);
+  }
+
+  clear(name) {
+    if (name) this.events.delete(name);
```

Comment:
I think this wouldn't clear the associated wrappers.

Reply:
This will require nested loops. See implemented in next commit but I'd prefer remove `ae.clear(name)` at all leaving clear all by `ae.clear()`.